routing/probability_bimodal.go

```diff
@@ -0,0 +1,238 @@
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+const (
+	// DefaultScaleMsat describes the distribution of funds in the LN. The
+	// value determines how quickly the bimodal distribution drops off from
+	// the edges of a channel. A larger value means, that the drop off is
+	// slow.
+	DefaultScaleMsat = lnwire.MilliSatoshi(300000000)
+)
+
+// probabilityBimodal returns node and pair probabilities based on historical
+// payment results based on a liquidity distribution model of the LN.
+type probabilityBimodal struct {
+	// ProbabilityEstimatorCfg contains configuration options for our
+	// estimator.
+	*ProbabilityEstimatorCfg
+}
+
+// Compile-time check that probabilityBimodal implements Estimator interface.
+var _ Estimator = (*probabilityBimodal)(nil)
+
+func (p *probabilityBimodal) getConfig() *ProbabilityEstimatorCfg {
+	return p.ProbabilityEstimatorCfg
+}
+
+func (p *probabilityBimodal) setConfig(cfg *ProbabilityEstimatorCfg) {
+	p.ProbabilityEstimatorCfg = cfg
+}
+
+// getPairProbability estimates the probability of successfully traversing to
+// toNode based on historical payment outcomes for the from node. Those outcomes
+// are passed in via the results parameter.
+func (p *probabilityBimodal) getPairProbability(now time.Time,
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity btcutil.Amount) float64 {
+
+	// We first compute the probability for the desired hop taking into
+	// account previous knowledge.
+	directProbability := p.getLiquidityProbability(
+		now, results, toNode, amt, lnwire.NewMSatFromSatoshis(capacity),
+	)
+
+	// The final probability is computed by taking into account other
+	// channels of the node.
+	return p.calculateProbability(directProbability, now, results, toNode)
+}
+
+// getLocalPairProbability computes the probability to reach toNode given a set
+// of previous learnings.
+func (p *probabilityBimodal) getLocalPairProbability(now time.Time,
+	results NodeResults, toNode route.Vertex) float64 {
+
+	return 1.0
+}
+
+// getLiquidityProbability computes the probability to reach a node based
+// on the liquidity distribution in the LN.
+func (p *probabilityBimodal) getLiquidityProbability(now time.Time,
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity lnwire.MilliSatoshi) float64 {
+
+	// Set success and failure amounts.
+
+	// We know that we can send zero satoshis.
+	successAmount := lnwire.MilliSatoshi(0)
+	// We know that we cannot send the full capacity.
+	failAmount := capacity
+
+	result, ok := results[toNode]
+	if ok {
+		if !result.FailTime.IsZero() {
+			// Apply time decay for the amount we can not send.
+			failAmount = cannotSend(
+				result.FailAmt,
+				capacity,
+				now,
+				result.FailTime,
+				p.PenaltyHalfLife,
+			)
+		}
+		if !result.SuccessTime.IsZero() {
+			// Apply time decay for the amount we can send.
+			successAmount = canSend(
```

Comment:
Same comment as below about decaying successes.

Reply:
See the comment above.